lib/src/Uncertainty/Bayesian/openturns/ThreeDVAR.hxx

```diff
@@ -0,0 +1,333 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#ifndef OPENTURNS_THREEDVAR_HXX
+#define OPENTURNS_THREEDVAR_HXX
+
+#include "openturns/OTprivate.hxx"
+#include "openturns/CalibrationAlgorithmImplementation.hxx"
+#include "openturns/Sample.hxx"
+#include "openturns/Function.hxx"
+#include "openturns/OptimizationAlgorithm.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/CenteredFiniteDifferenceGradient.hxx"
+
+BEGIN_NAMESPACE_OPENTURNS
+
+/**
+ * @class THREEDVAR
+ *
+ * @brief The class implements the 3D-VAR calibration
+ *
+ */
+class OT_API ThreeDVAR
+  : public CalibrationAlgorithmImplementation
+{
+  CLASSNAME
+public:
+
+  /** Default constructor */
+  ThreeDVAR();
+
+  /** Parameter constructor */
+  ThreeDVAR(const Function & model,
+	    const Sample & inputObservations,
+	    const Sample & outputObservations,
+	    const Point & candidate,
+	    const CovarianceMatrix & parameterCovariance,
+	    const CovarianceMatrix & errorCovariance);
+
+  /** String converter */
+  virtual String __repr__() const;
+
+  /** Performs the actual computation. Must be overloaded by the actual calibration algorithm */
+  virtual void run();
+  Point run(const Sample & inputObservations,
+	    const Sample & outputObservations,
+	    const Point & candidate,
+	    const TriangularMatrix & parameterInverseCholesky,
+	    const TriangularMatrix & errorInverseCholesky);
+
+  /** Algorithm accessor */
+  OptimizationAlgorithm getAlgorithm() const;
+  void setAlgorithm(const OptimizationAlgorithm & algorithm);
+
+  /** Candidate accessor */
+  Point getCandidate() const;
+
+  /** Parameter covariance accessor */
+  CovarianceMatrix getParameterCovariance() const;
+
+  /** Error covariance accessor */
+  CovarianceMatrix getErrorCovariance() const;
+
+  /** Bootstrap size accessor */
+  UnsignedInteger getBootstrapSize() const;
+  void setBootstrapSize(const UnsignedInteger bootstrapSize);
+
+  /* Here is the interface that all derived class must implement */
+
+  /** Virtual constructor */
+  virtual ThreeDVAR * clone() const;
+
+  /** Method save() stores the object through the StorageManager */
+  virtual void save(Advocate & adv) const;
+
+  /** Method load() reloads the object from the StorageManager */
+  virtual void load(Advocate & adv);
+
+private:
+
+  /* Model to calibrate */
+  Function model_;
+
+  /* The input observations */
+  Sample inputObservations_;
+
+  /* The optimization algorithm */
+  OptimizationAlgorithm algorithm_;
+
+  /* Number of bootstrap replica */
+  UnsignedInteger bootstrapSize_;
+
+  /* The error covariance */
+  CovarianceMatrix errorCovariance_;
+
+#ifndef SWIG
+  class CalibrationModelEvaluation: public EvaluationImplementation
```

Comment:
this should be moved to the .cxx, eventhough I cant see where its used

Reply:
@jschueller It is used in ThreeDVAR::run(), line 130. It allows to build the residual function of the 3DVAR optimization problem. Why do you want me to move this code to the .cxx? If defined as a nested class, it is automatically protected against name collision (eg. a collision with NonLinearLeastSquaresCalibration::CalibrationModelEvaluation), which is not the case if moved to the .cxx, no?

Reply:
I thought it would be better to not define the implementation in the header to avoid symbols being inlined in client lib, but your're right name collision is a pain, and if you want to avoid to separate declaration and definition, it can be wrked around by renaming to ThreeDVARCalibrationModelEvaluation or something, but its not so good either.